### DISCUSSION TITLE
SwitchMode11 - New SetOption for disabling toggle action at Double-Click

### DISCUSSION THREAD

**Discussion by Seestern93** [arendst/Tasmota#12713] [Ideas]
### PROBLEM DESCRIPTION
Summary:
Feature Request for new SetOptionXXX to disable TOGGLE event at DOUBLE click with SwitchMode11 (and 12) to use the DOUBLE event without triggering the TOGGLE event.

Detailed:
For handling a dimmable LED over a wall switch I am currently using the SwitchMode11 for toggling the light (1x click => switch1#state=2) and for changing the brightness (hold = switch1#state=4 -> "-"; click + hold = switch1#state=5 -> "+"). Now I got a second light (not-dimmable) which I want to command over the same switch, therefore I implemented double-click feature -> 2x click = switch1#state=8 -> toggling light 2.
My problem now is, that with each double-click, the toggle action (switch1#state=2) from the single-click is also sent. That means, that I have to filter out the toggle event which makes things more complicated and less respsonsive. I checked several methods how to do, but the only thing which worked for me is using a ruletimer with 1s to delay the "single-click" to check whether the "double-click" event was executed within that time. That means that single-click is now delayed by (a little bit more than) 1s. (with backlog it did not work due to the backlog queue is always FIFO). 

For completeness, here are my full rules:
(mem1 is used as "-" and "+" switch)
(mem2 is used to check whether double-click happened -> 0; or not -> 1; which is checked at end of Ruletimer1)
(cmnd/tasmsw01/event togglepower is used to toggle the second light)

```lua
Rule1:
on system#boot do
	backlog mem1 -; mem2 0;
endon 
on switch1#state=2 do
	backlog Ruletimer1 1; mem2 1
endon 
on switch1#state=4 do
	Dimmer %mem1%
endon 
on switch1#state=5 do
	mem1 +
endon 
on switch1#state=6 do
	mem1 -
endon 
on switch1#state=8 do
	backlog mem2 0; publish cmnd/tasmsw01/event togglepower
endon

Rule2:
on Rules#Timer=1 do
	event togglepower=%mem2%
endon
on event#togglepower=1 do
	power1 toggle
endon
```

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Shelly Dimmer 2
- [x] Tasmota binary firmware version number used: Tasmota 9.2.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: OTA
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
08:44:31 CMD: Backlog Template; Module; GPIO 255
08:44:31 MQT: stat/tasmshdim01/RESULT = {"NAME":"Shelly Dimmer 2","GPIO":[0,3200,0,3232,5568,5600,0,0,0,0,192,0,320,4736],"FLAG":0,"BASE":18}
08:44:31 MQT: stat/tasmshdim01/RESULT = {"Module":{"0":"Shelly Dimmer 2"}}
08:44:32 MQT: stat/tasmshdim01/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"3200":"Serial Tx"},"GPIO2":{"0":"None"},"GPIO3":{"3232":"Serial Rx"},"GPIO4":{"5568":"SHD Boot 0"},"GPIO5":{"5600":"SHD Reset"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"192":"Switch_n1"},"GPIO15":{"0":"None"},"GPIO16":{"320":"Led_i1"},"GPIO17":{"4736":"ADC Temp"}}}

```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
08:46:12 CMD: Backlog Rule1; Rule2; Rule3
08:46:12 MQT: stat/tasmshdim01/RESULT = {"Rule1":"ON","Once":"OFF","StopOnError":"OFF","Length":300,"Free":211,"Rules":"on system#boot do backlog mem1 -; mem2 0; endon on switch1#state=2 do backlog Ruletimer1 1; mem2 1 endon on switch1#state=4 do Dimmer %mem1% endon on switch1#state=5 do mem1 + endon on switch1#state=6 do mem1 - endon on switch1#state=8 do backlog mem2 0; publish cmnd/tasmsw01/event togglepower endon"}
08:46:12 MQT: stat/tasmshdim01/RESULT = {"Rule2":"ON","Once":"OFF","StopOnError":"OFF","Length":96,"Free":415,"Rules":"on Rules#Timer=1 do event togglepower=%mem2% ENDON on event#togglepower=1 do power1 toggle ENDON"}
08:46:12 MQT: stat/tasmshdim01/RESULT = {"Rule3":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}

```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
08:46:40 CMD: Status 0
08:46:40 MQT: stat/tasmshdim01/STATUS = {"Status":{"Module":0,"DeviceName":"tasmshdim01","FriendlyName":["tasmshdim01"],"Topic":"tasmshdim01","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":0,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"","SwitchMode":[11,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
08:46:40 MQT: stat/tasmshdim01/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"External System","Uptime":"0T12:25:16","StartupUTC":"2021-07-20T19:21:24","Sleep":50,"CfgHolder":4617,"BootCount":14,"BCResetTime":"2021-02-11T13:18:10","SaveCount":1664,"SaveAddress":"FA000"}}
08:46:41 MQT: stat/tasmshdim01/STATUS2 = {"StatusFWR":{"Version":"9.2.0(tasmota)","BuildDateTime":"2020-12-21T15:03:40","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"482/699"}}
08:46:41 MQT: stat/tasmshdim01/STATUS3 = {"StatusLOG":{"SerialLog":0,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Seestern-Homenet",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["8010800B","0505C8000100060000005A00000000000000","08800000","00006000","00000000"]}}
08:46:41 MQT: stat/tasmshdim01/STATUS4 = {"StatusMEM":{"ProgramSize":586,"Free":416,"Heap":24,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"15701C","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
08:46:41 MQT: stat/tasmshdim01/STATUS5 = {"StatusNET":{"Hostname":"tasmshdim01","IPAddress":"10.69.0.47","Gateway":"10.69.0.1","Subnetmask":"255.255.240.0","DNSServer":"10.69.0.1","Mac":"D8:BF:C0:19:FD:E2","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
08:46:41 MQT: stat/tasmshdim01/STATUS6 = {"StatusMQT":{"MqttHost":"10.69.0.10","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_19FDE2","MqttUser":"raspberryMQTT","MqttCount":5,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30}}
08:46:41 MQT: stat/tasmshdim01/STATUS7 = {"StatusTIM":{"UTC":"2021-07-21T07:46:41","Local":"2021-07-21T08:46:41","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+01:00","Sunrise":"05:10","Sunset":"20:42"}}
08:46:41 MQT: stat/tasmshdim01/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
08:46:41 MQT: stat/tasmshdim01/STATUS10 = {"StatusSNS":{"Time":"2021-07-21T08:46:41","Switch1":"OFF","ANALOG":{"Temperature":44.3},"ENERGY":{"TotalStartTime":"2021-02-11T13:18:10","Total":0.093,"Yesterday":0.093,"Today":0.000,"Power":0},"TempUnit":"C"}}
08:46:41 MQT: stat/tasmshdim01/STATUS11 = {"StatusSTS":{"Time":"2021-07-21T08:46:41","Uptime":"0T12:25:17","UptimeSec":44717,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":5,"POWER":"OFF","Dimmer":37,"Fade":"ON","Speed":4,"LedTable":"OFF","Wifi":{"AP":1,"SSId":"Seestern-Homenet","BSSId":"DC:39:6F:AB:C4:63","Channel":6,"RSSI":70,"Signal":-65,"LinkCount":8,"Downtime":"0T00:02:26"}}}


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
See steps to reproduce / expected behavior below

```

### TO REPRODUCE
_Steps to reproduce the behavior:_

1. Set SwitchMode to 11
2. Double-Click the Switch button (should be a momentary switch)
3. Switch#state=2 (=TOGGLE) is sent together with Switch#state=8 (=DOUBLE)

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

1. Set new SetOptionXXX to 1
2. Execute steps 1 + 2 to reproduce behavior
3. Only Switch#state=8 (=DOUBLE) is sent

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by wrightsonm** [comment]
I have found this exact issue too with a slightly different use case.
I am using an Athom SW012EU which has a single push button per relay channel.
I am connecting this device over wifi, mqtt & home assistant to a zigbee dimmer SAMOTECH SM309.

In the unlikely event of an mqtt / wifi failure, the tasmota switch shall revert to using the relay in a momentary mode to control the dimmer module using the push switch.  Momentary relay control is acheived using `PulseTime1 2`
When mqtt is available, the tasmota module will publish mqtt commands to control the zigbee dimmer.  The relay becomes disconnected and increases the life of the relay.

I have looked into controlling the zigbee dimmer with a push button using both BUTTON and SWITCH modes.
Advantage of the button mode is i could have single up to penta press and hold control.  Unfortunately this button does not publish multiple hold messages nor does it publish a button released message so it is not possible to control the dimmer using only 1 physical button.

In SWITCH mode the dimming functionality works as expected:
Short press turn on or turn off
Hold press dim up
Another hold press dim down
If mqtt disconnects, the switch gains control of the relay

Double press is where the problem is as it publishes both toggle and double press messages.

This is my configuration:


```
switchmode2 12
setoption32 5
setoption34 50
```

```
Rule2
on system#boot var1 0 ENDON
ON switch2#state=2 DO publish light/cmnd/POWER TOGGLE ENDON
ON switch2#state=4 DO publish light/cmnd/DIMMER %var1% ENDON
ON switch2#state=7 DO event toggling1=%var1%   ENDON
ON switch2#state=8 DO publish light/cmnd/POWER DOUBLE ENDON
ON event#toggling1<1 DO var1 1 ENDON
ON event#toggling1>0 DO var1 0 ENDON

Rule2 1

RULE3
on mqtt#disconnected do backlog switchstate2 0; rule2 0 endon
on mqtt#connected do backlog switchstate2 12; rule2 1 endon

Rule3 1
```

for info, Rule2 var1 does not initialize correctly on first use without a restart.  as var1 is only initialized on boot. but it recovers after the first dimming attempt.

```
18:32:04.850 CMD: Backlog Template; Module; GPIO 255
18:32:04.876 MQT: stat/tasmota_84D661/RESULT = {"NAME":"Athom SW012EU","GPIO":[576,289,0,160,225,161,0,0,0,224,288,0,0,0],"FLAG":0,"BASE":18}
18:32:04.959 MQT: stat/tasmota_84D661/RESULT = {"Module":{"0":"Athom SW012EU"}}
18:32:05.063 MQT: stat/tasmota_84D661/RESULT = {"GPIO0":{"576":"LedLink_i"},"GPIO1":{"289":"Led2"},"GPIO2":{"0":"None"},"GPIO3":{"160":"Switch1"},"GPIO4":{"225":"Relay2"},"GPIO5":{"161":"Switch2"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"224":"Relay1"},"GPIO14":{"288":"Led1"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```

```
18:33:04.707 CMD: Backlog Rule1; Rule2; Rule3
18:33:04.778 MQT: stat/tasmota_84D661/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
18:33:04.865 MQT: stat/tasmota_84D661/RESULT = {"Rule2":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":335,"Free":176,"Rules":"on system#boot var1 0 ENDON ON switch2#state=2 DO publish light/cmnd/POWER TOGGLE ENDON ON switch2#state=4 DO publish light/cmnd/DIMMER %var1% ENDON ON switch2#state=7 DO event toggling1=%var1%   ENDON ON switch2#state=8 DO publish light/cmnd/POWER DOUBLE ENDON ON event#toggling1<1 DO var1 1 ENDON ON event#toggling1>0 DO var1 0 ENDON"}}
18:33:04.964 MQT: stat/tasmota_84D661/RESULT = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":121,"Free":390,"Rules":"on mqtt#disconnected do backlog switchstate2 0; rule2 0 endon on mqtt#connected do backlog switchstate2 12; rule2 1 endon"}}
```

```
18:33:59.961 CMD: Status 0
18:33:59.969 MQT: stat/tasmota_84D661/STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota","Tasmota2"],"Topic":"tasmota_84D661","ButtonTopic":"0","Power":3,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,12,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
18:34:00.000 MQT: stat/tasmota_84D661/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:19:00","StartupUTC":"2021-08-22T17:14:59","Sleep":50,"CfgHolder":4617,"BootCount":9,"BCResetTime":"2021-08-22T15:53:53","SaveCount":42,"SaveAddress":"FA000"}}
18:34:00.036 MQT: stat/tasmota_84D661/STATUS2 = {"StatusFWR":{"Version":"9.5.0(tasmota)","BuildDateTime":"2021-06-17T08:26:35","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"357/699"}}
18:34:00.068 MQT: stat/tasmota_84D661/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["cisco",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","050532000100060000005A0A000000000000","00000080","00006000","00000000"]}}
18:34:00.109 MQT: stat/tasmota_84D661/STATUS4 = {"StatusMEM":{"ProgramSize":607,"Free":396,"Heap":27,"ProgramFlashSize":1024,"FlashSize":4096,"FlashChipId":"16405E","FlashFrequency":40,"FlashMode":3,"Features":["00000809","8FDAC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6"}}
18:34:00.144 MQT: stat/tasmota_84D661/STATUS5 = {"StatusNET":{"Hostname":"tasmota_84D661-5729","IPAddress":"192.168.xx.xx","Gateway":"192.168.xx.xx","Subnetmask":"255.255.255.0","DNSServer":"192.168.xx.xx","Mac":"xx:xx:xx:xx:xx","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
18:34:00.178 MQT: stat/tasmota_84D661/STATUS6 = {"StatusMQT":{"MqttHost":"mqtt.xx.xx","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_84D661","MqttUser":"xxxx","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
18:34:00.211 MQT: stat/tasmota_84D661/STATUS7 = {"StatusTIM":{"UTC":"2021-08-22T17:34:00","Local":"2021-08-22T18:34:00","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":"+01:00","Sunrise":"05:53","Sunset":"19:52"}}
18:34:00.245 MQT: stat/tasmota_84D661/STATUS10 = {"StatusSNS":{"Time":"2021-08-22T18:34:00","Switch1":"ON","Switch2":"OFF"}}
18:34:00.278 MQT: stat/tasmota_84D661/STATUS11 = {"StatusSTS":{"Time":"2021-08-22T18:34:00","Uptime":"0T00:19:01","UptimeSec":1141,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"ON","POWER2":"ON","Wifi":{"AP":1,"SSId":"xxxx","BSSId":"3A:E8:29:9A:C4:37","Channel":11,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":1,"Downtime":"0T00:00:03"}}}
```

On single press the console shows:
```
18:18:45.808 RUL: SWITCH2#STATE=2 performs "publish light/cmnd/POWER TOGGLE"
18:18:45.816 MQT: light/cmnd/POWER = TOGGLE
```

On double press the console shows:
```
18:18:09.554 RUL: SWITCH2#STATE=2 performs "publish light/cmnd/POWER TOGGLE"
18:18:09.564 MQT: light/cmnd/POWER = TOGGLE
18:18:09.814 RUL: SWITCH2#STATE=8 performs "publish light/cmnd/POWER DOUBLE"
18:18:09.825 MQT: light/cmnd/POWER = DOUBLE
```

On Hold the console shows:

```
18:19:19.996 RUL: SWITCH2#STATE=4 performs "publish light/cmnd/DIMMER 0"
18:19:20.001 MQT: light/cmnd/DIMMER = 0
18:19:20.195 RUL: SWITCH2#STATE=4 performs "publish light/cmnd/DIMMER 0"
18:19:20.204 MQT: light/cmnd/DIMMER = 0
...omitted duplicate dimmer = 0 messages for brevity
18:19:23.649 RUL: SWITCH2#STATE=7 performs "backlog event toggling1=0"
18:19:23.669 MQT: stat/tasmota_84D661/RESULT = {"Event":"Done"}
18:19:23.720 RUL: EVENT#TOGGLING1<1 performs "var1 1"
18:19:23.728 MQT: stat/tasmota_84D661/RESULT = {"Var1":"1"}
```

**Reply by wrightsonm** [reply]
Currently it is not really possible to use the double press functionality for anything other than turning the light fully on or fully off.

There are comments on this PR about the original design intent of the double press which appears to be to set the brightness to 100% rather than any other action.
[https://github.com/arendst/Tasmota/pull/7603](https://github.com/arendst/Tasmota/pull/7603)

**Reply by Seestern93** [reply]
Yes, that is exactly our problem (as @wrightsonm seems to have exactly the same root problem, that double press triggers TOGGLE event before DOUBLE event).
But from my understanding, it would be technically no problem to add a new SetOption to the code which allows us to change the behavior between standard (= "old") behavior, and "new" behavior which waits for the command to be single or multi press.
Or do I miss something?

**Reply by barbudor** [reply]
Switches do not support multipress. Only Press and hold.
Multipress is only a Button feature.
Beside `Switchmode 11` is a very special mode with a very specific behavior as it implements a repetitive HOLD.

It is more than "just a setoption" because Switches doesn't have at all the logic for multipress.
And Buttons doesn't have the logic for Switchmode 11

So there I see no way to achieve what you are looking for.

**Reply by Seestern93** [reply]
@barbudor Yes that's right. It seems that there is a little mixup between switch and button features 😁. Especially because from my perspective `Switchmode 11` seems to work best with buttons, not with switches.
I am already analyzing the code in the support_switch.ino with the intend to make at least a proof-of-concept change. Even though I am not the very best at programming C. My usual projects are written with Python 😆. 
We'll see if I am able to compile it 😀

**Reply by barbudor** [reply]
Switches and Buttons : I'm talking about the Tasmota software component, not the physical part.

While Tasmota's "Button" is meant to be used with a temporary push-to-make button, "Switch" can be used either with temporary buttons or rocker switches.

When using SwitchMode 11 with a Tasmota "Switch" the expected physical part is a push button.

**Reply by Seestern93** [reply]
But I hope you will agree that having a software function called switch which has 15 different modes and only two of them are made for buttons it is a little mix up ;-)
Nevertheless, I will try to add the requested feature locally and will give feedback if it worked and if yes, how I did it.
Then we can check again if it fits to the concept, or if it stays a "local" function :)

**Comment by Seestern93** [comment]
Update: New functionality Switch#Mode=10 added over #13007 which is triggered when switch is triggered only once within time set in `SetOption32`.
-> Instead of original proposal, this feature is always active and does not depend on a new SetOption (and therefore can be used individually for each Switch)
